lib/gc/Transforms/MergeAllocTickBased.cpp

```diff
@@ -30,6 +30,69 @@ using namespace special_ticks;
 /// and default memory space.
 static bool isMemRefTypeOk(MemRefType type) { return type.hasStaticShape(); }
 
+static inline int64_t getSizeInBytes(MemRefType &memType) {
+  // treat bool (i1) as 1 byte. It may not be true for all targets, but we at
+  // least have a large enough size for i1
+  int64_t size = memType.getElementTypeBitWidth() / 8;
+  size = (size > 0) ? size : 1;
+  for (auto v : memType.getShape()) {
+    size *= v;
+  }
+  return size;
+}
+
+static bool needsHoistOutOfParallelLoop(Operation *op) {
+  Operation *parent =
+      op->getParentWithTrait<OpTrait::AutomaticAllocationScope>();
+  if (parent && isa<scf::ForallOp>(parent)) {
+    // check if the current allocation is between the nested pfor, and use
+    // inside the inner parallel loop
+    SmallVector<Operation *, 4> parallelOpInCurBlock;
+    Block *curBlock = op->getBlock();
+    for (auto &curOp : curBlock->getOperations()) {
+      if (isa<scf::ForallOp>(curOp)) {
+        parallelOpInCurBlock.push_back(&curOp);
+      }
+    }
+
+    if (parallelOpInCurBlock.empty())
+      return false;
+
+    for (auto *use : op->getUsers()) {
+      for (auto *parallelOp : parallelOpInCurBlock) {
+        if (parallelOp->isAncestor(use)) {
+          return true;
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
+static bool isForallLoopBoundStatic(Operation *op) {
+  bool isStatic = true;
+  if (auto forallOp = dyn_cast<scf::ForallOp>(op)) {
+    OpBuilder builder{forallOp->getContext()};
+    SmallVector<Value> upperBounds = forallOp.getUpperBound(builder);
+    SmallVector<Value> lowerBounds = forallOp.getLowerBound(builder);
```

Comment:
It will create constant Op if the bounds are constant (they are originally stored in attr, if bounds are constant). I think using `getMixed***Bound()` should be a better choice. You can use `isa<Value>(....)` to check if it is constant.

Reply:
Good point, this will create constant Op, I can switch to use the function you mentioned.